File: KungFu Chess/PhysicsFactory.py
from Board import Board
from Physics import IdlePhysics, MovePhysics, JumpPhysics, RestPhysics, Physics


class PhysicsFactory:
    """Instantiate the correct *Physics* subclass for a given state."""

    def __init__(self, board: Board):
        self.board = board

    def create(self, start_cell, state_name: str, cfg) -> Physics:
        speed = cfg.get("speed_m_per_sec", 0.0)

        name_l = state_name.lower()
        if name_l == "move" or name_l.endswith("_move"):
            cls = MovePhysics
        elif name_l == "jump":
            cls = JumpPhysics
        elif name_l.endswith("rest") or name_l == "rest":
            cls = RestPhysics
        else:
            cls = IdlePhysics

        if cls is RestPhysics:
            duration_ms = cfg.get("duration_ms", 3000)
            return cls(start_cell, self.board, duration_ms)  # type: ignore[arg-type]

        return cls(start_cell, self.board, speed)

File: KFC_java/src/main/java/PhysicsFactory.java
import org.json.JSONObject;

public class PhysicsFactory {
    private final Board board;
    public PhysicsFactory(Board board){ this.board=board; }

    public Physics create(Moves.Pair startCell, String stateName, JSONObject cfg){
        double speed = cfg.optDouble("speed_m_per_sec", 0.0);
        Physics phys;
        String name = stateName.toLowerCase();
        if(name.equals("move")||name.endsWith("_move")){
            phys = new Physics.MovePhysics(board, speed);
        } else if(name.equals("jump")){
            phys = new Physics.JumpPhysics(board, cfg.optDouble("duration_ms",100)/1000.0);
        } else if(name.endsWith("rest")||name.equals("rest")){
            phys = new Physics.RestPhysics(board, cfg.optDouble("duration_ms",3000)/1000.0);
        } else {
            phys = new Physics.IdlePhysics(board);
        }
        // initialise position
        phys.startCell = startCell;
        phys.currPosM = board.cellToM(startCell);
        return phys;
    }
} 